common/src/main/java/me/lucko/luckperms/common/commands/group/ListGroups.java

```diff
@@ -53,15 +55,27 @@ public void execute(LuckPermsPlugin plugin, Sender sender, ArgumentList args, St
             return;
         }
 
-        Message.GROUPS_LIST.send(sender);
-        plugin.getGroupManager().getAll().values().stream()
-                .sorted((o1, o2) -> {
+        int page = args.getIntOrDefault(0, 1);
+        int pageIndex = page - 1;
+
+        List<Group> groups = plugin.getGroupManager().getAll().values().stream().sorted((o1, o2) -> {
                     int i = Integer.compare(o2.getWeight().orElse(0), o1.getWeight().orElse(0));
                     return i != 0 ? i : o1.getName().compareToIgnoreCase(o2.getName());
-                })
-                .forEach(group -> {
-                    List<String> tracks = plugin.getTrackManager().getAll().values().stream().filter(t -> t.containsGroup(group)).map(Track::getName).collect(Collectors.toList());
-                    Message.GROUPS_LIST_ENTRY.send(sender, group, group.getWeight().orElse(0), tracks);
-                });
+                }).collect(Collectors.toList());
+
+        List<List<Group>> pages = Iterators.divideIterable(groups, 7);
+
+        if (pageIndex < 0 || pageIndex >= pages.size()) {
+            page = 1;
+            pageIndex = 0;
+        }
+
+        Message.SEARCH_SHOWING_GROUPS.send(sender, page, pages.size(), groups.size());
+        Message.GROUPS_LIST.send(sender);
+
+        for (Group group : pages.get(pageIndex)) {
+            List<String> tracks = plugin.getTrackManager().getAll().values().stream().filter(t -> t.containsGroup(group)).map(Track::getName).collect(Collectors.toList());
```

Comment:
It would be good to cache the `plugin.getTrackManager().getAll().values()` collection outside of this for loop, as the `getAll` implementation creates a copy of the `ImmutableMap` each time it is called. But if you are not paying attention to performance, as this command is called once usually it is not necessary. Just a suggestion.

Reply:
Good point, thanks